[DOC-430] Document Kafka support (#309)
Macrometacorp/docs: docs/streams/kafka_on_gdn/fluent-bit-client.md

---
sidebar_position: 60
title: Fluent Bit Client
---

Fluent Bit is an open-source and multi-platform log processor tool that aims to be a generic Swiss knife for logs processing and distribution.

> Fluent Bit works as a logging pipeline chunked into logical parts like Inputs, Parsers, Filters and Outputs. Kafka client implementation able in Output part. It means Fluent bit produces records and sends to Kafka.

Reviewer: Review capitalization in this sentence. Also, use the serial comma (comma before the and).

Full Kafka plugin documentation can be found [here](https://docs.fluentbit.io/manual/pipeline/outputs/kafka)

> Kafka plugin is on top of `libdrkafka` [library](https://github.com/edenhill/librdkafka/).

Reviewer: ```suggestion
The Fluent Bit Kafka plugin is built on top of `libdrkafka` [library](https://github.com/edenhill/librdkafka/).
```

### Installation

> The correct installation was proven on Ubuntu OS. The bellow steps will show how to build and install Fluent Bit with correct plugins and modules.

Reviewer: ```suggestion
The correct installation was proven on Ubuntu OS. The below steps will show how to build and install Fluent Bit with correct plugins and modules.
```

Reviewer: What do you mean by "was proven"?

#### Prequisites

> Build Fluent Bit from scratch requires installing `git` and then clone a repository from [GitHub](https://github.com/fluent/fluent-bit) and checkout to the latest version.

Reviewer: ```suggestion
Building Fluent Bit from scratch requires installing `git` and then cloning a repository from [GitHub](https://github.com/fluent/fluent-bit) and checkout to the latest version.
```

```
sudo apt update
sudo apt install git
git clone https://github.com/fluent/fluent-bit.git
git checkout 1.8.0-pack-fixes
```

> Building of Fluent Bit requires other Ubuntu packages.

Reviewer: What if I'm not using Ubuntu?

Author: I have made changes in doc

Fluent Bit uses CMake as it build system. The suggested procedure to prepare the build system consists on the following steps:

```
sudo apt install -y build-essential
sudo apt install -y make
sudo apt install -y cmake
sudo apt install -y libsasl2-dev
sudo apt install -y bison
sudo apt install -y flex
sudo apt install -y libssl-dev
```

#### Consumer

To verify that our message has indeed been produced to the topic, let's create a consumer to consume our message:

```node
const consumer = kafka.consumer({ groupId: '<my-group-id>' })

await consumer.connect()
await consumer.subscribe({ topic: '<c8globals or c8locals>.<my-topic>', fromBeginning: true })

await consumer.run({
  eachMessage: async ({ topic, partition, message }) => {
    console.log({
      value: message.value.toString(),
    })
  }
})
```

> ### Build

Reviewer: What are we building?

Author: The github project we got from https://github.com/fluent/fluent-bit.git

Change to the `build` directory inside the Fluent Bit sources and let CMake configure the project specifying where the root path is located with Kafka plugin and TLS module:

```
cd fluent-bit/build
cmake -DFLB_OUT_KAFKA=On -DFLB_TLS=On ..
make
```

> ### Run

Reviewer: This topic reads differently than previous topics. Different tone, different content. Try to be consistent with the different topics in this folder as much as possible. This helps readers situate themselves in the content.

Author: Fluent Bit works differently than other clients, hence the documentation is different

This sample uses on input `dummy` plugin. It generates records that consists event time. Here is an example `{"@timestamp":1629229398.754061,"message":"dummy"}`.
Output parameter uses `kafka` plugin with name of Kafka topic and additional parameters from `libdrkafka` library.
A topic should have prefix of global either local namespace. A full topic name: `<c8globals or c8locals>.<my-topic>`.

```
bin/fluent-bit \
    -i dummy \
    -o kafka \
    -p brokers=<my-gdn>.<my-paas>.macrometa.io:9093 \
    -p topics=<c8globals or c8locals>.<my-topic> \
    -p rdkafka.security.protocol=SASL_SSL \
    -p rdkafka.enable.ssl.certificate.verification=false \
    -p rdkafka.ssl.ca.location=<path-to-ca.cert.pem-file> \
    -p rdkafka.sasl.mechanism=PLAIN \
    -p rdkafka.sasl.username=<my-tenant>/<my-fabric> \
    -p rdkafka.sasl.password=token:<my-api-key or my-JWT> \
    -p rdkafka.client.id=fluent-bit-client \
    -p rdkafka.log.connection.close=false \
    -p rdkafka.request.required.acks=1
```

Note: The full list of properties can be found [here](https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md).

### Verify

> Verifying the correct sending messages can be performed via consumer of any java, python, node.js Kafka client. The list of clients can be found [here](README.md).

Reviewer 2: ```suggestion
Sent messages can be verified by the consumer of any java, python, node.js Kafka client. The list of clients can be found [here](README.md).
```
